packaging/metadata.py

```diff
@@ -0,0 +1,394 @@
+    from typing import Literal
+
+    NormalizedDynamicFields = Literal[
+        "platform",
+        "summary",
+        "description",
+        "keywords",
+        "home-page",
+        "author",
+        "author-email",
+        "license",
+        "supported-platform",
+        "download-url",
+        "classifier",
+        "maintainer",
+        "maintainer-email",
+        "requires-dist",
+        "requires-python",
+        "requires-external",
+        "project-url",
+        "provides-extra",
+        "provides-dist",
+        "obsoletes-dist",
+        "description-content-type",
+    ]
+else:
+    NormalizedDynamicFields = str
+
+
+def _normalize_field_name_for_dynamic(field: str) -> NormalizedDynamicFields:
+    """Normalize a metadata field name that is acceptable in `dynamic`.
+
+    The field name will be normalized to lower-case. JSON field names are
+    also acceptable and will be translated accordingly.
+
+    """
+    return cast(NormalizedDynamicFields, field.lower().replace("_", "-"))
+
+
+# Bypass frozen dataclass for __post_init__, this approach is documented in:
+# https://docs.python.org/3/library/dataclasses.html#frozen-instances
+_setattr = object.__setattr__
+
+
+# In the following we use `frozen` to prevent inconsistencies, specially with `dynamic`.
+# Comparison is disabled because currently `Requirement` objects are
+# unhashable/not-comparable.
+
+
+@dataclasses.dataclass(frozen=True, eq=False)
+class CoreMetadata:
+    """
+    Core metadata for Python packages, represented as an immutable
+    :obj:`dataclass <dataclasses>`.
+
+    Specification: https://packaging.python.org/en/latest/specifications/core-metadata/
+
+    Attribute names follow :pep:`PEP 566's JSON guidelines
+    <566#json-compatible-metadata>`.
+    """
+
+    # 1.0
+    name: str
+    version: Union[Version, None] = None
+    platform: Collection[str] = ()
+    summary: str = ""
+    description: str = ""
+    keywords: Collection[str] = ()
+    home_page: str = ""
+    author: str = ""
+    author_email: Collection[Tuple[Union[str, None], str]] = ()
+    license: str = ""
+    # license_file: Collection[str] = ()  # not standard yet
+    # 1.1
+    supported_platform: Collection[str] = ()
+    download_url: str = ""
+    classifier: Collection[str] = ()
+    # 1.2
+    maintainer: str = ""
+    maintainer_email: Collection[Tuple[Union[str, None], str]] = ()
+    requires_dist: Collection[Requirement] = ()
+    requires_python: SpecifierSet = dataclasses.field(default_factory=SpecifierSet)
+    requires_external: Collection[str] = ()
+    project_url: Mapping[str, str] = dataclasses.field(default_factory=dict)
+    provides_extra: Collection[str] = ()
+    provides_dist: Collection[Requirement] = ()
+    obsoletes_dist: Collection[Requirement] = ()
+    # 2.1
+    description_content_type: str = ""
+    # 2.2
+    dynamic: Collection[NormalizedDynamicFields] = ()
+
+    def __post_init__(self) -> None:
```

Comment:
I don't know if this method is necessary. If you mess up on what you assign then that's on you; language of consenting adults and all. Otherwise you should be running a type checker to make sure you are keeping your types consistent with what people expect.

Reply:
Thank you very much for the comment @brettcannon. Here I tried to follow the principle of being loose with the input but strict with the output. But I completely understand your point.

Something that I am also thinking now is: since `packaging` aims to be a collection of "reusable core utilities for various Python Packaging interoperability specifications", it would be nice if it could handle some of the tedious and repetitive processing that would be required to use this class (e.g. converting strings to `Requirement` or `SpecifierSet` objects) and are currently implemented in `__post_init__`.

So what if I remove this method as you suggested, but add another method that would handle this use case? For example a `from__unstructured_dict` class method (or any other name really, please feel free to bikeshed) with proper and explicit type annotations?

(Please note that my intention here is not to force the types to be correct, but instead absorb the repetitive tasks that would have to be implemented by any backend using such metadata API)

Reply:
Thank you very much for the comment @pradyunsg! I will bring this discussion to #383 then 😄

Reply:
I agree with @pradyunsg : the structure of the data that's going to be coming in is very much going to be from known formats. The only one that might be different is someone trying to convert `setup.cfg`/`setup.py`, but that doesn't need to live here (i.e. we don't have to be all things to all people).

Reply:
I think such a method would be useful to implement `from_pyproject` so I thought that making it public would allow people making more use of the class even before `from_pyproject` gets implemented.

But I have no problems with the proposed approach. I will trust in your intuition here.

Reply:
@abravalheri it's more paranoia. 😅 It's much harder to take something away then to introduce something to an API. As such, we prefer to keep the APIs small and for scenarios we know exist rather than guess at future needs. "Premature optimization" applies to API design as well. 😉